Allow cantrip deck creation from spell

KSops opened the pull request:
Resolves #2256

KSops commented:
I'm sure if there's demand for "other items" we can address it then.
As for demand for the deck, there's at least the original ticket author as well as myself. Maybe I'm playing my class wrong :P but I keep a few decks on my rogue for utility.

CarlosFdez commented:
I know right now you can simulate them by creating scrolls and adding uses to them. I guess this is a pretty unique case where its "pick any spell" I suppose.

Sure then. But for now might be best to lock it to first level spells.

CarlosFdez commented on src/module/actor/sheet/base.ts:
```diff
@@ -719,7 +726,7 @@ export abstract class ActorSheetPF2e<TActor extends ActorPF2e> extends ActorShee
                     {},
                     async (heightenedLevel, itemType, spell) => {
-                        if (!(itemType === "scroll" || itemType === "wand")) return;
+                        if (!isSpellConsumableItemType(itemType)) return;
```
You should probably remove this check and make the function's type into SpellConsumableItemType.

CarlosFdez commented on src/module/item/consumable/spell-consumables.ts:
```diff
@@ -31,26 +36,50 @@ const wandCompendiumIds: Record<number, string | undefined> = {
+        case "scroll":
+            name = game.i18n.format("PF2E.ScrollFromSpell", { name: spellName, level: heightenedLevel });
+            break;
+        default:
```
Don't use default for this. Use "wand". That way if anything is ever added, typescript will notify you that a change is required.

CarlosFdez replied:
Even better, have a mapping from item type to localization entry. Then pull from that mapping.

CarlosFdez commented:
You'll need to find a way to hide level selection when it changes to cantrip. You could look at the spellcasting create and edit dialog for implementation advice.

Instead of a new isSpellConsumable function, use tupleHasValue() or setHasElement() and have the value list somewhere. MAGIC_TRADITIONS can be searched as an example.

stwlam commented:
> I'm sure if there's demand for "other items" we can address it then. As for demand for the deck, there's at least the original ticket author as well as myself. Maybe I'm playing my class wrong :P but I keep a few decks on my rogue for utility.

Demand for this is slight compared to "other items," and I'm not convinced it's worth hard-coding.

CarlosFdez commented:
The only reason I'm kinda entertaining it is because its basically just a different more restricted version of a scroll. This shouldn't happen too often, but ideally if we're gonna hardcode it, it'd be nice if it was cleaned up first.

That said, perhaps a better way of doing this is to add dropzones to scroll and wand details that show the embedded spell. That way the cantrip deck can just be an empty "scroll".

KSops commented:
Emulating this with empty scrolls affects a character's wealth (sure, that's editable but it's work), carried bulk, and missing item description about how cantrip decks work.

But if the decision is that request itself is not worth it then the original ticket should be marked as `won't fix` and we should close it accordingly with whatever justifications.

CarlosFdez commented:
I think what I plan to do is the dropzone once I find a bit more time (perhaps after I add custom damage type support), and we'll see how well that handles your use case. If it doesn't handle it well, we can consider adding additional options to this menu (after its cleaned up of course). I don't think its problematic to add dialog options for certain common cases so long as there is a non-hardcoded way to do it in the system.

KSops commented:
I was going over the rules and noticed cantrips can't be loaded into scrolls or wands. I'm thinking of leaving the wand scroll popup as it originally was then and adding a basic confirm dialog for creation of a cantrip. tldr; if it's a cantrip being dragged and dropped, switch to a confirm dialog that creates a cantrip deck, else go to original behaviour. Does that sound sensible?

CarlosFdez commented:
Thinking more on it, after a bunch of weeks, I don't disagree with the concept. I think the idea is fine. While we need to implement dropzones anyways for these sorts of items, this dialog is more of a user friendliness thing than a new feature persay.

That said, creating a whole new dialog type for just cantrips concerns me a bit. I would suggest renaming ScrollWandPopup to CastingItemCreateDialog and do it there if possible. Try to reduce the code to a minimum, so its not a pain to add more if we ever need to.

For the localizations, rather than creating new top level ones, it might be best to bunch them up. My recommendation is `PF2E.Item.Physical.FromSpell.X` where X is Cantrip5, Scroll, and Wand. Shark might ask you to change this later, he tends to have different suggestions on where to put these things. Feel free to remove the old localization entries for this.

There might eventually be more but that's what comes to mind. This code is part of the not-cleaned-up part of the codebase, so the end result should likely be less code than what you started with (or about par). Adding more is a sign that something probably went wrong.

KSops commented:
- Changed dialogs back into a single dialog
- Renamed dialog as suggested
- Changed localization strings
    - Wands are apparently not consumables so some strings were changed to call them "spellcasting items"
    - Reordered some strings in `en.json` to be sorted alphanumerically
    - Grouped item creation names as suggested

CarlosFdez commented on src/module/item/consumable/spell-consumables.ts:
```diff
@@ -31,26 +41,36 @@ const wandCompendiumIds: Record<number, string | undefined> = {
+        default:
+            id = wandCompendiumIds[heightenedLevel] ?? null;
+    }
+    return id;
```
return directly from inside the case statements rather than assigning to a let. You could also alternatively chain a ternary.

In general try to avoid lets whenever possible.

CarlosFdez commented on src/module/actor/sheet/popups/casting-item-create-dialog.ts:
```diff
@@ -0,0 +1,107 @@
+    onSubmitCallback: CastingItemCreateCallback;
+    spell: SpellPF2e;
+    formDataCache: FormOutputData;
+    isCantrip: boolean;
```
replacing all internal isCantrip references with this.spell.isCantrip will save you a few lines. There's no real need to cache that.

KSops replied:
Just thought I'd save lookups but I guess it comes from a set, so it'll be negligible

CarlosFdez commented on static/templates/popups/casting-item-create-dialog.hbs:
```diff
@@ -0,0 +1,26 @@
+        <select name="level" data-dtype="Number">
+        {{#select level}}
+            {{#each validLevels}}
+            <option value="{{this}}">{{this}}</option>
```
you should probably match what heightenLevels does in spell-overlay.hbs

KSops replied:
Which part are you referring to? Using a named local instead of `this`? Or a localization string?

CarlosFdez replied:
I mean ordinal values like 3rd Level, 4th Level. But I'm looking at the current dialog again, and this isn't a regression, so its ok to leave as is.

CarlosFdez commented on src/module/actor/sheet/popups/casting-item-create-dialog.ts:
```diff
@@ -0,0 +1,105 @@
+        return options;
+    }
+
+    override async getData(): Promise<FormApplicationData<ActorPF2e>> {
```
You should make the FormInputData the return type, and have it extend from FormApplicationData<ActorPF2e>. Rather than setting properties on the super, you should return a new object with the old properties splatted at the start. Most sheet classes more or less follow that pattern now.

CarlosFdez commented on src/module/actor/sheet/popups/casting-item-create-dialog.ts:
```diff
@@ -0,0 +1,105 @@
+            await this.formSubmit(event, formData);
+        } else {
+            await this.formUpdate(event, formData);
+        }
```
I don't think this method and the ones called after are big enough to need to split it up like this. I think what you can do is that event.type !== "submit", you do an await this.render() and then return. And then you do what's currently in formSubmit() afterwards.

KSops replied:
Sure I can do that. Just thought I'd make it readable and clear so that someone can change behaviour in the future for a specific course of action.

CarlosFdez commented:
I'm ok with this as written now. Thanks for working on it.
